File: apps/api/src/services/notification.service.ts
import type { HydratedDocument } from "mongoose";
import {
  DEFAULT_PAGE_SIZE,
  MAX_PAGE_SIZE,
  type Notification,
  type NotificationType,
  type PaginatedResult,
} from "@believe-ai/shared";
import type { NotificationDocument } from "@believe-ai/server";
import { notificationRepository } from "../repositories/notification.repository.js";
import { logger } from "../config/logger.js";
import { NotFoundError } from "../errors/AppError.js";

function toDto(doc: HydratedDocument<NotificationDocument>): Notification {
  return {
    id: doc._id.toString(),
    userId: doc.userId.toString(),
    type: doc.type as NotificationType,
    title: doc.title,
    body: doc.body ?? "",
    link: doc.link ?? null,
    read: doc.read,
    createdAt: doc.createdAt.toISOString(),
  };
}

export interface CreateNotificationInput {
  type: NotificationType;
  title: string;
  body: string;
  link?: string | null;
}

export const notificationService = {
  /**
   * Like audit writes, a failed notification must never break the action
   * that triggered it — an import shouldn't 500 because the bell couldn't
   * be updated.
   */
  async create(userId: string, input: CreateNotificationInput): Promise<void> {
    try {
      await notificationRepository.create({
        userId,
        type: input.type,
        title: input.title,
        body: input.body,
        link: input.link ?? null,
      });
    } catch (err) {
      logger.error({ err, type: input.type }, "Failed to create notification");
    }
  },

  async list(
    userId: string,
    page = 1,
    limit = DEFAULT_PAGE_SIZE,
  ): Promise<PaginatedResult<Notification> & { unread: number }> {
    const safePage = Math.max(1, page);
    const safeLimit = Math.min(MAX_PAGE_SIZE, Math.max(1, limit));
    const [[items, total], unread] = await Promise.all([
      notificationRepository.list(userId, safePage, safeLimit),
      notificationRepository.countUnread(userId),
    ]);

    return {
      items: items.map(toDto),
      page: safePage,
      limit: safeLimit,
      total,
      totalPages: Math.max(1, Math.ceil(total / safeLimit)),
      unread,
    };
  },

  async markRead(id: string, userId: string): Promise<Notification> {
    const doc = await notificationRepository.markRead(id, userId);
    if (!doc) throw new NotFoundError("Notification not found");
    return toDto(doc);
  },

  async markAllRead(userId: string): Promise<number> {
    const result = await notificationRepository.markAllRead(userId);
    return result.modifiedCount;
  },
};

File: apps/backend/services/notification_service.py
"""In-app notifications — reads/marks-read the caller's own notifications;
create() is exposed for future ports of the actions that trigger them
(see repositories/notification_repository.py)."""

from __future__ import annotations

from bson import ObjectId

from core.errors import NotFoundError
from core.logging import get_logger
from models.notification import Notification, NotificationType
from repositories import notification_repository
from schemas.notification import NotificationDto, NotificationListResult
from schemas.pagination import DEFAULT_PAGE_SIZE, safe_limit, safe_page, total_pages

logger = get_logger(__name__)


def _to_dto(doc: Notification) -> NotificationDto:
    return NotificationDto(
        id=str(doc.id),
        userId=str(doc.userId),
        type=doc.type,
        title=doc.title,
        body=doc.body,
        link=doc.link,
        read=doc.read,
        createdAt=doc.createdAt.isoformat(),
    )


async def create(user_id: ObjectId, type_: NotificationType, title: str, body: str, link: str | None = None) -> None:
    """A failed notification must never break the action that triggered it —
    same contract as Node's notificationService.create."""
    try:
        await notification_repository.create(user_id, type_, title, body, link)
    except Exception as err:  # noqa: BLE001 — deliberately broad, see docstring
        logger.error("Failed to create notification of type %s: %s", type_, err)


async def list_notifications(
    user_id: ObjectId, page: int = 1, limit: int = DEFAULT_PAGE_SIZE
) -> NotificationListResult:
    safe_page_ = safe_page(page)
    safe_limit_ = safe_limit(limit)
    items, total = await notification_repository.list_for_user(user_id, safe_page_, safe_limit_)
    unread = await notification_repository.count_unread(user_id)
    return NotificationListResult(
        items=[_to_dto(doc) for doc in items],
        page=safe_page_,
        limit=safe_limit_,
        total=total,
        totalPages=total_pages(total, safe_limit_),
        unread=unread,
    )


async def mark_read(notification_id: ObjectId, user_id: ObjectId) -> NotificationDto:
    doc = await notification_repository.mark_read(notification_id, user_id)
    if not doc:
        raise NotFoundError("Notification not found")
    return _to_dto(doc)


async def mark_all_read(user_id: ObjectId) -> int:
    return await notification_repository.mark_all_read(user_id)
